jsettlers.logic/src/jsettlers/ai/highlevel/AiStatistics.java

```diff
@@ -0,0 +1,612 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ *******************************************************************************/
+package jsettlers.ai.highlevel;
+
+import static jsettlers.common.buildings.EBuildingType.FARM;
+import static jsettlers.common.buildings.EBuildingType.LUMBERJACK;
+import static jsettlers.common.buildings.EBuildingType.WINEGROWER;
+import static jsettlers.common.landscape.ELandscapeType.RIVER1;
+import static jsettlers.common.landscape.ELandscapeType.RIVER2;
+import static jsettlers.common.landscape.ELandscapeType.RIVER3;
+import static jsettlers.common.landscape.ELandscapeType.RIVER4;
+import static jsettlers.common.mapobject.EMapObjectType.STONE;
+import static jsettlers.common.mapobject.EMapObjectType.TREE_ADULT;
+import static jsettlers.common.movable.EMovableType.SWORDSMAN_L1;
+import static jsettlers.common.movable.EMovableType.SWORDSMAN_L2;
+import static jsettlers.common.movable.EMovableType.SWORDSMAN_L3;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+
+import jsettlers.algorithms.construction.AbstractConstructionMarkableMap;
+import jsettlers.common.CommonConstants;
+import jsettlers.common.buildings.EBuildingType;
+import jsettlers.common.buildings.IBuilding;
+import jsettlers.common.landscape.ELandscapeType;
+import jsettlers.common.landscape.EResourceType;
+import jsettlers.common.map.shapes.MapCircle;
+import jsettlers.common.mapobject.EMapObjectType;
+import jsettlers.common.material.EMaterialType;
+import jsettlers.common.movable.EMovableType;
+import jsettlers.common.position.ShortPoint2D;
+import jsettlers.logic.buildings.Building;
+import jsettlers.logic.map.grid.MainGrid;
+import jsettlers.logic.map.grid.flags.FlagsGrid;
+import jsettlers.logic.map.grid.landscape.LandscapeGrid;
+import jsettlers.logic.map.grid.movable.MovableGrid;
+import jsettlers.logic.map.grid.objects.ObjectsGrid;
+import jsettlers.logic.map.grid.partition.PartitionsGrid;
+import jsettlers.logic.movable.Movable;
+import jsettlers.logic.objects.stack.StackMapObject;
+import jsettlers.logic.player.Player;
+
+/**
+ * This class calculates statistics based on the grids which are used by highlevel and lowlevel KI. The statistics are calculated once and read
+ * multiple times within one AiExecutor step triggerd by the game clock.
+ * 
+ * @author codingberlin
+ */
+public class AiStatistics {
+
+	private static final short BORDER_LAND_WIDTH = 10;
+
+	private final Queue<Building> buildings;
+	private Map<Integer, Map<EBuildingType, Integer>> totalBuildingsNumbers;
```

Comment:
I introduced PlayerStatistic class. It's instances are stored in an array.
It is not called AiPlayerStatistic because it holds statistics from human players as well.
This is needed later when the ai decides to attack or not to attack by the number of soldiers of the human player.